File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/ordering/OrderLine.java
package com.zappymart.domain.ordering;

import com.zappymart.domain.shared.Money;

import java.util.Objects;

public record OrderLine(String productId, String productName, Money unitPrice, int quantity) {

    public OrderLine {
        Objects.requireNonNull(productId, "productId");
        Objects.requireNonNull(productName, "productName");
        Objects.requireNonNull(unitPrice, "unitPrice");
        if (quantity < 1) {
            throw new IllegalArgumentException("An order line holds at least one product: " + quantity);
        }
    }

    public Money lineTotal() {
        return unitPrice.times(quantity);
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/ordering/OrderLine.kt
package nl.zappymart.domain.ordering

import nl.zappymart.domain.shared.Money

data class OrderLine(
    val productId: String,
    val productName: String,
    val unitPrice: Money,
    val quantity: Int,
) {
    val lineTotal: Money get() = unitPrice * quantity
}
